bot/common/threads/onboarding.py

```diff
@@ -98,56 +118,315 @@ async def handle_emoji(self, raw_reaction):
+
+    name = StepKeys.ADD_USER_WALLET_ADDRESS.value
+
+    def __init__(self, user_id, guild_id, cache):
+        super().__init__(user_id, guild_id, cache)
+
+    def get_account_descriptor(self):
+        return "Ethereum wallet address"
+
+    def sanitize_account_input(self, message):
+        stripped_message = message.content.strip()
+        if not Web3.isAddress(stripped_message):
+            raise ThreadTerminatingException(
+                "Supplied address %s is not a valid ethereum address" % stripped_message
+            )
+
+        return stripped_message
+
+    def get_account_storage_key(self):
+        return WALLET_STORAGE_KEY
 
     async def handle_emoji(self, raw_reaction):
         return _handle_skip_emoji(raw_reaction, self.guild_id)
 
 
+class PromptUserWalletMessageSignatureStep(PromptForAccountVerificationStep):
+    """Prompts the user to sign a message with their specified wallet"""
+
+    name = StepKeys.PROMPT_USER_WALLET_ADDRESS.value
+
+    def __init__(self, user_id, guild_id, cache):
+        super().__init__(user_id, guild_id, cache)
+
+    def get_account_verification_prompts(self):
+        instructions_embed = discord.Embed(
+            colour=INFO_EMBED_COLOR,
+            title="Verification instructions",
+            description=WALLET_VERIFICATION_INSTRUCTIONS,
+        )
+
+        prompt = discord.Embed(
+            colour=INFO_EMBED_COLOR,
+            title="Please sign the below message with your wallet",
+            description=REQUESTED_SIGNED_MESSAGE,
+        )
+        return [instructions_embed, prompt]
+
+
+class VerifyUserWalletMessageSignatureStep(VerifyAccountStep):
+    """Verifies the message supplied by the user was signed by their specified wallet"""
+
+    name = StepKeys.VERIFY_USER_WALLET_ADDRESS.value
+
+    def __init__(self, user_id, guild_id, cache):
+        super().__init__(user_id, guild_id, cache)
+
+    async def verify_account(self, authentication_message):
+        address = await get_cache_metadata(self.user_id, self.cache, WALLET_STORAGE_KEY)
+        stripped_supplied_signature = authentication_message.content.strip()
+
+        try:
+            int(stripped_supplied_signature)
+        except ValueError:
+            raise ThreadTerminatingException("The response wasn't a correct signature!")
+
+        requested_msg_hex = encode_defunct(text=REQUESTED_SIGNED_MESSAGE)
+        recovered_address = Account.recover_message(
+            requested_msg_hex, signature="0x" + stripped_supplied_signature
+        )
+
+        if address != recovered_address:
+            raise ThreadTerminatingException(
+                "Recovered address from message & signature doesn't match supplied"
+            )
+
+    async def save_authenticated_account(self):
+        #  retrieve and save handle from cache into airtable
+        record_id = await find_user(self.user_id, self.guild_id)
+        #  TODO: should this be .toLowered?
```

Comment:
Yes, I think in python it is .lower()

Reply:
yep, i just meant it colloquially. but you agree that we should just have our wallet addresses stored as all-lowercase?